### PR TITLE
31078 misp

### PR DESCRIPTION
**Pull request by maimorag**
## Status
- [x] In Progress

## Related Issues
fixes: https://jira-dc.paloaltonetworks.com/browse/XSUP-31078

## Description
Adding a limit to fetch indicators command

### CODE REVIEW

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 1]
```diff
@@ -583,6 +583,7 @@ def main():
     timeout = arg_to_number(params.get('timeout')) or 60
     insecure = not params.get('insecure', False)
     proxy = params.get('proxy', False)
+    fetch_limit = int(params.get('limit', '2000'))
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:586]:
```suggestion
    fetch_limit = arg_to_number(params.get('limit', '2000'))
```

File: Packs/FeedMISP/ReleaseNotes/1_0_31.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+
+#### Integrations
+
+##### MISP Feed
+- Adding a **limit** to ***fetch-indicators***.
```
Comment (ShirleyDenkberg) [inline Packs/FeedMISP/ReleaseNotes/1_0_31.md:5]:
```suggestion
- Adding a **limit** argument to the ***fetch-indicators*** command.
```

**Comment by ShirleyDenkberg** [comment]
@eyalpalo Doc review completed.

File: Packs/FeedMISP/ReleaseNotes/1_0_31.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+
+#### Integrations
+
+##### MISP Feed
+- Adding a **page** argument to the ***misp-feed-get-indicators*** command.
+- Added pagination to the ***fetch-indicators*** command.
```
Comment (eyalpalo) [inline Packs/FeedMISP/ReleaseNotes/1_0_31.md:6]:
```suggestion
- Fixed an issue where ***fetch-indicators*** crushed due to memory overflow.
```

**Review by eyalpalo — CHANGES_REQUESTED** [review]

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 4]
```diff
@@ -558,23 +541,52 @@ def get_attributes_command(client: Client, args: Dict[str, str], params: Dict[st
...
+        timestamp = search_query_per_page.get("response", {}).get("Attribute", [])[-1].get('timestamp')
+        params_dict['page'] += 1
+        search_query_per_page = client.search_query(params_dict)
+    if error_message := search_query_per_page.get('Error'):
+        raise DemistoException(f"Error in API call - check the input parameters and the API Key. Error: {error_message}")
+    params_dict.pop("limit", None)
+    params_dict.pop("page", None)
+    demisto.setLastRun({
+        'params': params_dict,
+        'timestamp': timestamp})
```
[the remark this answers is not in the document]
Comment (maimorag) [reply Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:589]: removed

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 5]
```diff
@@ -558,23 +541,52 @@ def get_attributes_command(client: Client, args: Dict[str, str], params: Dict[st
...
     attribute_types = argToList(params.get('attribute_types', ''))
     query = params.get('query', None)
-    indicators = fetch_indicators(client, tags, attribute_types, query, tlp_color,
-                                  params.get('url'), reputation, feed_tags)
-    return indicators
+    params_dict = clean_user_query(query) if query else build_params_dict(tags, attribute_types)
+    params_dict['page'] = 1
+    params_dict['limit'] = 2000
+    timestamp = ''
+    search_query_per_page = client.search_query(params_dict)
```
[the remark this answers is not in the document]
Comment (maimorag) [reply Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:562]: removed duplicates

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 6]
```diff
@@ -558,23 +526,37 @@ def get_attributes_command(client: Client, args: Dict[str, str], params: Dict[st
...
     reputation = params.get('feedReputation')
     tags = argToList(params.get('attribute_tags', ''))
     feed_tags = argToList(params.get("feedTags", []))
     attribute_types = argToList(params.get('attribute_types', ''))
     query = params.get('query', None)
-    indicators = fetch_indicators(client, tags, attribute_types, query, tlp_color,
-                                  params.get('url'), reputation, feed_tags)
-    return indicators
+    params_dict = clean_user_query(query) if query else build_params_dict(tags=tags, attribute_type=attribute_types, limit=2000,
+                                                                          page=1)
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:545]: Are we sure we should start at page 1 and not at page 0?

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 7]
```diff
@@ -558,23 +526,37 @@ def get_attributes_command(client: Client, args: Dict[str, str], params: Dict[st
...
+                        page: {params_dict["page"]}')
+        indicators = build_indicators(search_query_per_page, attribute_types, tlp_color, params.get('url'), reputation, feed_tags)
+        demisto.createIndicators(indicators)
+        params_dict['page'] += 1
+        search_query_per_page = client.search_query(params_dict)
+    if error_message := search_query_per_page.get('Error'):
+        raise DemistoException(f"Error in API call - check the input parameters and the API Key. Error: {error_message}")
+    params_dict.pop("limit", None)
+    params_dict.pop("page", None)
+    demisto.setLastRun({'params': params_dict})
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:559]: Do we use the last run anywhere in the code? If not, lets remove it

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.yml [hunk 8]
```diff
@@ -127,6 +127,9 @@ script:
     - name: limit
       defaultValue: '10'
       description: The maximum number of results to return.
+    - name: page
+      defaultValue: '0'
+      description: The maximum number of results to return.
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.yml:132]: The description is wrong. Lets update it.

**Review by eyalpalo — CHANGES_REQUESTED** [review]

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py (hunk 6 above)
Comment (maimorag) [reply Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:545]: According to docs it stating at page 1
https://www.misp-project.org/openapi/#tag/Attributes/operation/restSearchAttributes

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py (hunk 7 above)
Comment (maimorag) [reply Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:559]: ok, removing

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py (hunk 6 above)
Comment (maimorag) [reply Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:516]: I also fixed the default value

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 9]
```diff
@@ -537,8 +471,13 @@ def get_attributes_command(client: Client, args: Dict[str, str], params: Dict[st
     feed_tags = argToList(params.get("feedTags", []))
     query = args.get('query', None)
     attribute_type = argToList(args.get('attribute_type', ''))
-    indicators = fetch_indicators(client, tags, attribute_type,
-                                  query, tlp_color, params.get('url'), reputation, feed_tags, limit, False)
+    page = arg_to_number(args.get('page', '1')) or 0
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:474]:
```suggestion
    page = arg_to_number(args.get('page')) or 1
```

File: Packs/FeedMISP/ReleaseNotes/1_0_31.md [hunk 10]
```diff
@@ -0,0 +1,7 @@
+
+#### Integrations
+
+##### MISP Feed
+- Adding a **page** argument to the ***misp-feed-get-indicators*** command.
```
Comment (eyalpalo) [inline Packs/FeedMISP/ReleaseNotes/1_0_31.md:5]:
```suggestion
- Added a **page** argument to the ***misp-feed-get-indicators*** command.
```

**Review by eyalpalo — APPROVED** [review]

**Pull request merged**